### PR TITLE
CI: start testing Python 3.11

### PR DESCRIPTION
**Pull request by fangchenli**
- [ ] xref #46680 (Replace xxxx with the Github issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by mroeschke** [comment]
Is this going to supersede https://github.com/pandas-dev/pandas/pull/47032?

**Comment by fangchenli** [comment]
> 

Didn't realize there is an existing PR...

**Comment by lithomas1** [comment]
You can just push to my PR directly if you want, or pull the changes here. It's a little farther down the pipeline, but I will not have time to work on it in the near future.

**Comment by fangchenli** [comment]
```
FAILED pandas/tests/arrays/categorical/test_api.py::TestCategoricalAPI::test_set_ordered
FAILED pandas/tests/arrays/categorical/test_api.py::TestPrivateCategoricalAPI::test_codes_immutable
FAILED pandas/tests/indexes/multi/test_get_set.py::test_set_levels_codes_directly
FAILED pandas/tests/indexes/period/test_freq_attr.py::TestFreq::test_freq_setter_deprecated
FAILED pandas/tests/io/parser/test_quoting.py::test_null_quote_char[python--0]
FAILED pandas/tests/io/parser/test_quoting.py::test_null_quote_char[python--3]
FAILED pandas/tests/io/parser/common/test_read_errors.py::test_null_byte_char[python]
ERROR pandas/tests/scalar/test_nat.py - TypeError: '<' not supported between ...
ERROR pandas/tests/scalar/test_nat.py - TypeError: '<' not supported between ...
```

**Comment by lithomas1** [comment]
@fangchenli 
Thanks for picking this up. There haven't been commits on this PR in a while, so I was wondering whether you had the time to finish this PR up.

I've merged main into this PR for you just to see where we're at with the failures.

I'm pretty close to finishing off #47467, and am going to try to get it merged this week. I think I have enough time where I'll be able to finish it up, hopefully before 3.11rc1.

**Comment by lithomas1** [comment]
I accidentally pushed some commits here(I misconfigured github cli-meant to push to my own branch and pull request to this one), so I decided to go all the way to getting CI green. Sorry about 

At any rate, this is ready now pending #47467 merge.

File: .github/workflows/python-dev.yml [hunk 1]
```diff
@@ -1,11 +1,3 @@
-# This file is purposely frozen(does not run). DO NOT DELETE IT
```
Comment (lithomas1) [inline .github/workflows/python-dev.yml:1]: I think it would be better to leave the comment in, and just comment out the ``if: false`` below.

Comment (mroeschke) [reply .github/workflows/python-dev.yml:1]: Likewise. Even better to clarify

1. "Freeze" if our dependencies don't support the latest python dev version
2. "Unfreeze" if our dependencies support it, but the python dev version hasn't been officially released

File: pandas/tests/io/parser/common/test_read_errors.py [hunk 2]
```diff
@@ -224,13 +225,19 @@ def test_read_csv_wrong_num_columns(all_parsers):
...
     data = "\x00,foo"
     names = ["a", "b"]
     parser = all_parsers
 
-    if parser.engine == "c":
+    if parser.engine == "c" or (parser.engine == "python" and PY311):
+        if parser.engine == "python" and PY311:
+            request.node.add_marker(
+                pytest.mark.xfail(
+                    reason="In Python 3.11, this is read as an empty character not null"
```
Comment (mroeschke) [inline pandas/tests/io/parser/common/test_read_errors.py:238]: Curious, is there a documented change on the python side that notes this?

File: .github/workflows/python-dev.yml (hunk 1 above)
Comment (lithomas1) [reply .github/workflows/python-dev.yml:1]: @fangchenli Do you have any objections here?

File: pandas/tests/io/parser/common/test_read_errors.py (hunk 2 above)
Comment (lithomas1) [reply pandas/tests/io/parser/common/test_read_errors.py:238]: I can't find anything in the whatsnew right now.

Comment (mroeschke) [reply pandas/tests/io/parser/common/test_read_errors.py:238]: Might be worth raising and issue on the cpython issue tracker to see if this is expected. Would save some triage work if pandas users encounter this behavior change.

Comment (lithomas1) [reply pandas/tests/io/parser/common/test_read_errors.py:238]: Nvm, it's actually intended. https://github.com/python/cpython/pull/28808

We can just leave it xfailed for now, and I (or someone else) can patch the Python parser to take this into account.

File: .github/workflows/python-dev.yml (hunk 1 above)
Comment (fangchenli) [reply .github/workflows/python-dev.yml:1]: > @fangchenli Do you have any objections here?

No, I agree with you and @mroeschke

File: pandas/tests/io/parser/common/test_read_errors.py (hunk 2 above)
Comment (mroeschke) [reply pandas/tests/io/parser/common/test_read_errors.py:238]: Okay sounds good. Agreed to leave this xfailed and addressed later.

**Review by mroeschke — COMMENTED** [review]
Looks good. Just need to address https://github.com/pandas-dev/pandas/pull/47442#discussion_r943726182

**Review by mroeschke — APPROVED** [review]
LGTM pending green

**Comment by fangchenli** [comment]
@lithomas1 Thank you very much for finishing this up.

**Comment by lithomas1** [comment]
> @lithomas1 Thank you very much for finishing this up.

Thanks for picking this up again.

**Pull request merged**